lib/jackalope/persistent_work_list.ex

```diff
@@ -0,0 +1,357 @@
+defmodule Jackalope.PersistentWorkList do
+  @moduledoc """
+  A genserver wrapper for CubQ which we leverage to store and restore worklist tasks during disconnections
+  """
+  use GenServer
+
+  alias Jackalope.WorkList.Expiration
+
+  require Logger
+
+  @default_max_size 100
+
+  @tick_delay 10 * 60 * 1_000
+
+  defmodule State do
+    @moduledoc false
+    defstruct db: nil,
+              queue: nil,
+              queue_name: nil,
+              max_work_list_size: nil,
+              expiration_fn: nil,
+              update_expiration_fn: nil
+  end
+
+  @doc "Create a new work list"
+  @spec new(function(), function(), non_neg_integer(), Keyword.t()) :: pid()
+  def new(expiration_fn, update_expiration_fn, max_size \\ @default_max_size, opts \\ []) do
+    args = [
+      expiration_fn: expiration_fn,
+      update_expiration_fn: update_expiration_fn,
+      max_size: max_size,
+      opts: opts
+    ]
+
+    Logger.info("[Jackalope] Starting #{__MODULE__} with #{inspect(opts)}")
+    {:ok, pid} = GenServer.start_link(__MODULE__, args)
+    pid
+  end
+
+  @impl GenServer
+  def init(args) do
+    opts = Keyword.fetch!(args, :opts)
+    data_dir = Keyword.get(opts, :data_dir, "/data/jackalope")
+    queue_name = Keyword.get(opts, :queue_name, :items)
+
+    db =
+      case CubDB.start_link(data_dir: data_dir, name: :work_list, auto_compact: true) do
+        {:ok, pid} ->
+          pid
+
+        {:error, {:already_started, pid}} ->
+          pid
+
+        other ->
+          Logger.warn("[Jackalope] Corrupted DB : #{inspect(other)}. Erasing it.")
+          _ = File.rmdir(data_dir)
+          raise "Corrupted work list DB"
+      end
```

Comment:
How about something like this? The GenServer should never return already_started. It would be nice to more specifically know that the DB is corrupt rather than always erasing, but I didn't look to see if that was possible. Crashing after fixing the corruption just to get `init` called again seemed like it might have ramifications on sibling GenServers. I didn't want to think about documenting the behavior if the caller used `:all_for_one`.


```suggestion
    cubdb_opts = [data_dir: data_dir, name: :work_list, auto_compact: true]
    {:ok, db} =
      case CubDB.start_link(cubdb_opts) do
        {:error, reason} ->
          Logger.warn("[Jackalope] Corrupted DB : #{inspect(reason)}. Erasing it.")
          _ = File.rmdir(data_dir)
          CubDB.start_link(cubdb_opts)
          
         success -> success
      end
```

Reply:
What if start_link fails for reasons other than DB corruption?

Reply:
The same thing as before is fine. I.e., crash.

Reply:
Committed and pushed